### PR TITLE
feat: add Dropout layer to TensorMap model builder

### PR DESCRIPTION
**Pull request by Sickwoman**
## Description
This PR adds support for a Dropout layer in the TensorMap model builder.

## Changes
- Added DropoutNode component
- Added Dropout layer to the sidebar
- Integrated DropoutNode into the ReactFlow canvas

## Testing
- Verified that Dropout layer appears in the sidebar
- Dragged Dropout node into the canvas
- Connected it with other layers successfully

## Screenshots
Dropout layer available in the model builder sidebar and canvas.
Screenshot from 2026-03-04 20-29-39

### CODE REVIEW

**Comment by rajutkarsh07** [comment]
fix the lints too and this branch is out of date with the base branch so fix this too

File: tensormap-frontend/src/components/DragAndDropCanvas/Sidebar.jsx [hunk 1]
```diff
@@ -11,35 +12,49 @@ function Sidebar() {
...
         </div>
+
+        <div
+          className="cursor-grab rounded-md border border-l-4 border-l-node-dense bg-white px-3 py-2 text-xs font-medium"
+          onDragStart={(e) => onDragStart(e, "customdropout")}
+          draggable
+        >
+          Dropout
+        </div>
+
```
Comment (ivantha) [inline tensormap-frontend/src/components/DragAndDropCanvas/Sidebar.jsx]: **[MEDIUM]** Dropout Sidebar Item Reuses `border-l-node-dense` Color — Visual Ambiguity

The Dropout layer in the sidebar uses the same CSS color class as the Dense layer (`border-l-node-dense`):

```jsx
className="cursor-grab rounded-md border border-l-4 border-l-node-dense bg-white px-3 py-2 text-xs font-medium"
```

Looking at the pattern across all other nodes:
- `border-l-node-input` → Input
- `border-l-node-dense` → Dense
- `border-l-node-flatten` → Flatten
- `border-l-node-conv` → Conv2D

Each layer type has a distinct color for quick visual identification. Reusing `border-l-node-dense` for Dropout makes it visually indistinguishable from Dense in the sidebar, which is confusing and inconsistent. Notably, the `DropoutNode` canvas component also uses generic white styling without a color accent.

**Suggestion:**
```
Define a new Tailwind theme color for Dropout (e.g., in `tailwind.config.js`):
```js
// tailwind.config.js
theme: {
  extend: {
    colors: {
      'node-dropout': '#a78bfa', // purple-400, for example
    }
  }
}
```
Then use it in Sidebar.jsx:
```jsx
className="cursor-grab rounded-md border border-l-4 border-l-node-dropout bg-white px-3 py-2 text-xs font-medium"
```
And apply a matching border accent to `DropoutNode.jsx`:
```jsx
<div className="bg-white border-l-4 border-l-node-dropout border rounded shadow px-3 py-2 text-xs">
```
```

File: tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+import PropTypes from "prop-types";
+
+function DropoutNode({ data }) {
+  return (
+    <div className="bg-white border rounded shadow px-3 py-2 text-xs">
+
+      <Handle type="target" position={Position.Left} />
+
+      <div className="font-semibold">Dropout</div>
+      <div>Rate: {data?.rate ?? 0.5}</div>
```
Comment (ivantha) [inline tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx]: **[MEDIUM]** Dropout Rate Is Read-Only — Users Cannot Configure It

The Dropout rate is displayed statically:
```jsx
<div>Rate: {data?.rate ?? 0.5}</div>
```

The default of `0.5` is hardcoded in the display, and there is no input control to let users change the rate. In a neural network model builder, the dropout rate is a critical hyperparameter (valid range: 0.0–1.0) that users must be able to configure.

Other nodes in the codebase (Dense, Conv2D) almost certainly have editable fields for their hyperparameters (units, kernel size, etc.). This node is missing that core functionality, making it incomplete as a usable layer. Additionally, if the rate is never set on `data`, the model JSON generator downstream will likely use `null` or undefined for the rate rather than the visual fallback of `0.5`, potentially generating an invalid model configuration.

**Suggestion:**
```
Add an editable input for the rate, following the pattern of other custom nodes. For example:
```jsx
import { Handle, Position } from "reactflow";
import PropTypes from "prop-types";

function DropoutNode({ data, id }) {
  const handleRateChange = (e) => {
    const newRate = parseFloat(e.target.value);
    if (!isNaN(newRate) && newRate >= 0 && newRate < 1) {
      data.onChange?.(id, { rate: newRate });
    }
  };

  return (
    <div className="bg-white border-l-4 border-l-node-dropout border rounded shadow px-3 py-2 text-xs">
      <Handle type="target" position={Position.Left} />
      <div className="font-semibold mb-1">Dropout</div>
      <label className="flex items-center gap-1">
        Rate:
        <input
          type="number"
          min={0}
          max={0.99}
          step={0.01}
          defaultValue={data?.rate ?? 0.5}
          onChange={handleRateChange}
          className="w-14 border rounded px-1"
        />
      </label>
      <Handle type="source" position={Position.Right} />
    </div>
  );
}
```
```

File: tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+      <Handle type="source" position={Position.Right} />
+
+    </div>
+  );
+}
+
+DropoutNode.propTypes = {
+  data: PropTypes.object
+};
```
Comment (ivantha) [inline tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx]: **[MEDIUM]** PropTypes Definition Is Too Loose

```js
DropoutNode.propTypes = {
  data: PropTypes.object
};
```

Using `PropTypes.object` provides no information about the expected shape of the `data` prop. It:
1. Gives no type safety or tooling guidance for the `rate` field.
2. Does not mark `data` as required, meaning a missing `data` prop won't trigger a warning (though `data?.rate` handles the runtime case).
3. Does not validate the `rate` field's type or range.

In addition, ReactFlow custom nodes also receive `id`, `selected`, and `xPos`/`yPos` props which are not declared here at all.

**Suggestion:**
```
Use `PropTypes.shape` for a more descriptive definition:
```js
DropoutNode.propTypes = {
  id: PropTypes.string.isRequired,
  selected: PropTypes.bool,
  data: PropTypes.shape({
    rate: PropTypes.number,
    onChange: PropTypes.func,
  }),
};
```
```

File: tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+function DropoutNode({ data }) {
+  return (
+    <div className="bg-white border rounded shadow px-3 py-2 text-xs">
+
+      <Handle type="target" position={Position.Left} />
+
+      <div className="font-semibold">Dropout</div>
+      <div>Rate: {data?.rate ?? 0.5}</div>
+
+      <Handle type="source" position={Position.Right} />
```
Comment (ivantha) [inline tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx]: **[LOW]** Handles Missing Explicit IDs — May Cause Connection Issues

Both `Handle` components lack an `id` attribute:
```jsx
<Handle type="target" position={Position.Left} />
<Handle type="source" position={Position.Right} />
```

While ReactFlow allows handles without explicit IDs when there is only one handle of each type, explicitly providing IDs is a best practice for future-proofing. If a second handle of the same type is ever added (e.g., a skip connection), the lack of IDs will cause ambiguity. Additionally, if the `generateModelJSON` helper inspects handle IDs to build the graph topology, missing IDs could cause silent bugs when connecting nodes.

**Suggestion:**
```
```jsx
<Handle type="target" position={Position.Left} id="input" />
<Handle type="source" position={Position.Right} id="output" />
```
```

File: tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    </div>
+  );
+}
+
+DropoutNode.propTypes = {
+  data: PropTypes.object
+};
+
+export default DropoutNode;
```
Comment (ivantha) [inline tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx]: **[LOW]** No Unit Tests for DropoutNode Component

The PR introduces a new `DropoutNode` component but includes no corresponding test file. Looking at the PR's own testing section, it only mentions manual UI testing (drag-and-drop).

Key behaviors that should be tested:
- Renders with a provided rate value
- Renders with fallback rate of 0.5 when `data.rate` is undefined
- Renders with `data` being undefined (full optional path)
- Handles are rendered in the correct positions
- The node renders without errors when `data` is null/undefined

**Suggestion:**
```
Create `DropoutNode.test.jsx` alongside the component:
```jsx
import { render, screen } from "@testing-library/react";
import DropoutNode from "./DropoutNode";
import { ReactFlowProvider } from "reactflow";

const wrapper = ({ children }) => <ReactFlowProvider>{children}</ReactFlowProvider>;

describe("DropoutNode", () => {
  it("renders with a specified rate", () => {
    render(<DropoutNode data={{ rate: 0.25 }} />, { wrapper });
    expect(screen.getByText("Rate: 0.25")).toBeInTheDocument();
  });

  it("falls back to rate 0.5 when rate is not provided", () => {
    render(<DropoutNode data={{}} />, { wrapper });
    expect(screen.getByText("Rate: 0.5")).toBeInTheDocument();
  });

  it("renders without crashing when data is undefined", () => {
    render(<DropoutNode />, { wrapper });
    expect(screen.getByText("Dropout")).toBeInTheDocument();
  });
});
```
```

File: tensormap-frontend/src/components/DragAndDropCanvas/Sidebar.jsx (hunk 1 above)
Comment (ivantha) [inline tensormap-frontend/src/components/DragAndDropCanvas/Sidebar.jsx]: **[LOW]** Spurious Blank Lines Added Throughout Sidebar.jsx

The diff shows multiple blank lines added inside JSX and function bodies that were not there before:
- A blank line after `function Sidebar() {` (line 4)
- Blank lines between each draggable `<div>` block
- A blank line after `<CardContent ...>` (line 17)
- A blank line after the Dropout `</div>` before `</CardContent>`

These blank lines are not harmful, but they are unnecessary noise in the diff. They make the PR harder to review (the diff is larger than it needs to be), mix style changes with functional changes, and diverge from the original file's formatting conventions without apparent reason.

**Suggestion:**
```
Revert the spurious blank lines to keep the diff focused on the intentional changes (adding the Dropout sidebar entry). Only include whitespace changes if they were intentional and purposeful.
```

**Review by ivantha — COMMENTED** [review]
## ◎ Argus Review

This PR successfully adds a Dropout layer to the TensorMap model builder, following existing patterns for node registration and sidebar integration. However, it includes several notable issues: hardcoded database credentials committed to version control (critical security concern), a broken import path alias, a missing persistent DB volume, and a Dropout node that lacks an editable rate property and visual distinction. The core ReactFlow integration is clean and consistent with the existing codebase patterns.

File: tensormap-backend/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+services:
+  db:
+    image: postgres:16
+    environment:
+      POSTGRES_PASSWORD: mysecretpassword
+      POSTGRES_USER: postgres
+      POSTGRES_DB: mydatabase
```
Comment (ivantha) [inline tensormap-backend/docker-compose.yml]: **[LOW]** No Port Mapping or Restart Policy Defined

The docker-compose service is missing:
1. **Port mapping**: Without `ports: - "5432:5432"`, the backend application cannot connect to the database from the host or from other services (unless they are in the same Docker network).
2. **Restart policy**: Without `restart: unless-stopped`, the container will not automatically recover from crashes or system reboots during development.

**Suggestion:**
```
```yaml
services:
  db:
    image: postgres:16
    restart: unless-stopped
    ports:
      - "5432:5432"
    environment:
      POSTGRES_PASSWORD: ${POSTGRES_PASSWORD}
      POSTGRES_USER: ${POSTGRES_USER:-postgres}
      POSTGRES_DB: ${POSTGRES_DB:-mydatabase}
    volumes:
      - postgres_data:/var/lib/postgresql/data

volumes:
  postgres_data:
```
```

File: tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx (hunk 3 above)
Comment (Sickwoman) [reply tensormap-frontend/src/components/DragAndDropCanvas/CustomNodes/DropoutNode/DropoutNode.jsx]: I've updated the PropTypes definition to use `PropTypes.shape` and explicitly defined the expected structure for `data` (including `rate` and `onChange`), along with the `id` and `selected` props received from ReactFlow.

File: tensormap-frontend/src/components/DragAndDropCanvas/Canvas.jsx [hunk 7]
```diff
@@ -42,6 +43,7 @@ const nodeTypes = {
   customdense: DenseNode,
   customflatten: FlattenNode,
   customconv: ConvNode,
+  customdropout: DropoutNode,
 };
```
[the remark this answers is not in the document]
Comment (Sickwoman) [reply tensormap-frontend/src/components/DragAndDropCanvas/Canvas.jsx]: Thanks for the feedback! Glad to hear the DropoutNode integration follows the existing node registration pattern. I tried to keep the changes minimal and consistent with the other custom nodes.

**Comment by ivantha** [comment]
This branch has merge conflicts with `main`. Please rebase onto the latest `main`, resolve the conflicts, and push an update.
